File: src/backend/studio-reservation/src/main/java/org/taonaw/studio_reservation/domain/model/shared/TimeRange.java
package org.taonaw.studio_reservation.domain.model.shared;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;
import org.taonaw.studio_reservation.domain.shared.Assertion;

import java.time.LocalDate;
import java.time.LocalTime;

@Getter
@EqualsAndHashCode
public class TimeRange {
    private final LocalTime startTime;
    private final LocalTime endTime;

    public TimeRange(LocalTime startTime, LocalTime endTime) {
        Assertion.required(startTime);
        Assertion.required(endTime);
        this.startTime = startTime;
        this.endTime = endTime;
    }

    public boolean hasStartTimeSeconds() {
        return startTime.getSecond() != 0 || startTime.getNano() != 0;
    }

    public boolean hasEndTimeSeconds() {
        return endTime.getSecond() != 0 || endTime.getNano() != 0;
    }

    public boolean isStraddleTheDay() {
        return !startTime.isBefore(endTime);
    }

    public DateTimeRange toDateTimeRange(@NonNull LocalDate date) {
        var startDateTime = date.atTime(getStartTime());
        var endDateTime = isStraddleTheDay() ? date.plusDays(1).atTime(getEndTime()) : date.atTime(getEndTime());
        return new DateTimeRange(startDateTime, endDateTime);
    }

    public boolean isOverlapped(@NonNull TimeRange other) {
        var sampleDate = LocalDate.now();

        var thisStartDateTime = sampleDate.atTime(this.startTime);
        var thisEndDateTime = this.isStraddleTheDay() ?
                sampleDate.plusDays(1).atTime(this.endTime) :
                sampleDate.atTime(this.endTime);

        var otherStartDateTime = sampleDate.atTime(other.startTime);
        var otherEndDateTime = other.isStraddleTheDay() ?
                sampleDate.plusDays(1).atTime(other.endTime) :
                sampleDate.atTime(other.endTime);

        return thisStartDateTime.isBefore(otherEndDateTime) && otherStartDateTime.isBefore(thisEndDateTime);
    }
}

File: src/backend/reservation/src/main/kotlin/org/taonaw/reservation/domain/model/shared/TimeRange.kt
package org.taonaw.reservation.domain.model.shared

import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime

interface TimeRange {
    val start: LocalTime
    val end: LocalTime

    fun hasStartTimeSeconds(): Boolean {
        return start.second != 0 || start.nano != 0
    }

    fun hasEndTimeSeconds(): Boolean {
        return end.second != 0 || end.nano != 0
    }

    fun isStraddleTheDay(): Boolean {
        return !start.isBefore(end)
    }

    fun isOverlapped(other: TimeRange): Boolean {
        val sampleDate = LocalDate.now()

        val thisStartDateTime = sampleDate.atTime(start)
        val thisEndDateTime = if (isStraddleTheDay()) {
            sampleDate.plusDays(1).atTime(end)
        } else {
            sampleDate.atTime(end)
        }

        val otherStartDateTime = sampleDate.atTime(other.start)
        val otherEndDateTime = if (other.isStraddleTheDay()) {
            sampleDate.plusDays(1).atTime(other.end)
        } else {
            sampleDate.atTime(other.end)
        }

        return thisStartDateTime.isBefore(otherEndDateTime) && otherStartDateTime.isBefore(thisEndDateTime)
    }

    fun toDateTimeRange(date: LocalDate): DateTimeRange {
        val startDateTime: LocalDateTime = date.atTime(start)
        val endDateTime: LocalDateTime = if (isStraddleTheDay()) {
            date.plusDays(1).atTime(end)
        } else {
            date.atTime(end)
        }

        return BasicDateTimeRange(startDateTime, endDateTime)
    }

    override fun equals(other: Any?): Boolean
    override fun hashCode(): Int
    override fun toString(): String
}

data class BasicTimeRange(
        override val start: LocalTime,
        override val end: LocalTime) : TimeRange {
}